### DISCUSSION TITLE
Orchard Core as an Intranet

### DISCUSSION THREAD

**Discussion by kevinasdzine** [OrchardCMS/OrchardCore#16987] [Q&A]
I'm prototyping an Intranet using Orchard Core with Blazor wasm and I was wondering if there are any other contributors using the platform this way and anything they can share?

I think the basic idea of having no content types accessible anonymously covers a bunch (i.e. just accessing the site forces a login). All users would have access to some content types in the Admin - e.g. "creating Suggestion content items" and might be able to create some other types as Draft only.

However, I'd also like the published content to be interactive through the Theme to some point (e.g. liking or voting on content). I'm assuming I would just tie additional SQL tables back to different content instances in Documents to track this sort of thing. However, before I "roll my own" I'd like to see if there is any existing examples of this.

Anyone else running their intranet on Orchard Core?

**Comment by Piedone** [comment]
With Blazor WASM there shouldn't be anything special compared to any frontend framework. As far as OC is concerned, probably the most complex part will be authenticating with the web APIs.

While we at Lombiq have worked and work on quite a few functionally intranet websites (not intranet in the network sense, but apps that are accessible solely for people within a company, despite hosted on the internet), since those are private, I'm not at a liberty of talking about them, but if you have specific questions, I'll try to answer.

**Reply by kevinasdzine** [reply]
I wasn't looking to make a pure Blazor front-end (I still like the shape templates for most content presentation) but just for places where I want enhance the client experience - e.g. forms with cascading behaviors.

I'm curious about the use of additional SQL tables in parallel with Orchard's document database (like tracking "likes" on an article). I believe a migration can still add arbitrary tables to the database. Can they be related to documents?

**Reply by Piedone** [reply]
Yes and yes. You can do this completely independent of OC too, with any ORM or plain SQL. Better is to utilize at least migrations indeed, but even if not, `IDbConnectionAccessor` to get the current connection instead of opening a new one.

We do this with custom tables and LINQ to DB BTW, see https://github.com/Lombiq/Helpful-Libraries/tree/dev/Lombiq.HelpfulLibraries.LinqToDb.

OC tables are just tables as usual, however, you need to take care of the relationships with documents and other tables (e.g. when the document is deleted, the row in your custom table should also be deleted). While you can attempt this with custom foreign keys, the easiest approach for custom but document-related tables is to just create YesSql index tables, because those are managed automatically.

**Reply by kevinasdzine** [reply]
examples of CreateMapIndexTableAsync I've seen have an `IndexProvider` that updates the fields in the index from the fields in the content item. However, I'm trying to create a one to many relationship between the content item and the user's votes (each user gets one vote). So, I'm assuming I will need a separate table with a foreign key to the content item and a foreign key to the User (unless that is wrapped up in one of these `SchemaBuilder` methods already and I just haven't found an example of it).

**Reply by kevinasdzine** [reply]
It's a bit difficult to follow the code but would AuditTrail be a good model for this sort of data?

**Reply by Piedone** [reply]
Yes, you'll need to build some kind of relationship for this. Or, alternatively, you can store votes in a single document too, depending on what data you want to keep (like if it's just a couple of key value pairs for voting options and number of votes, then you can keep the whole thing in a document).

While I'm sure one can learn a lot from it, I don't think Audit Trail would be a particularly good example for this, although it does have relationships. From the top of my head, I can't think of an example of something really matching. For managing non-content item entities in general, check out the `Book` example in https://github.com/Lombiq/Orchard-Training-Demo-Module.

**Reply by kevinasdzine** [reply]
The Book example was appropriate for this use-case, thanks.

**Comment by mariojsnunes** [comment]
I've been doing that for a few years, but with Angular.
We don't allow any user access the Orchard backoffice, instead we provide a custom UX to create/manage the Content.
This is our website if you want to peek: https://diggspace.com/

**Reply by kevinasdzine** [reply]
I'm having a little trouble prototyping a "vote" for a content item. I think I just need to define an object with the ContentItemId and a UserId, create an index provider and then use `SchemaBuilder.CreateMapIndexTableAsync` in my migration but it's not working quite like I'm expecting. Can you point me to any example code?

**Reply by mariojsnunes** [reply]
I'd recommend creating a custom table for that.
Those MapIndexes are useful when you want to query/aggregate properties of a content item. Votes should be a separate thing, not exist inside of a content item.

**Reply by kevinasdzine** [reply]
Would that still be accessed via the ISession interface? For example

```
var articleVotes = await _session<ArticleVote, ArticleVoteIndex>(index => index.ContentItemId == thisArticle.ContentItemId).ListAsync();
```

**Reply by mariojsnunes** [reply]
no, you could use IDbConnectionAccessor (with raw sql) or some ORM to query your sql server directly.
you can still use SchemaBuilder to create the table via a Migration:

```
await SchemaBuilder.CreateTableAsync<Vote>();
await SchemaBuilder.AlterTableAsync(nameof(Vote), table => table.CreateIndex(...));
```

**Reply by kevinasdzine** [reply]
Interesting. I'm still trying to figure out what @Piedone was referring to about creating index tables to do this type of thing. I'd like to stay as "orchard centric" as possible (I had another OC project where I paralleled a separate EF Core set of tables but it was a bit cumbersome)

**Reply by mariojsnunes** [reply]
you could use an index if you stored all the votes on an article content item, but you don't want to do that as it could cause many document updates and a huge document size. In my experience, for some features you are better off with custom code.

**Reply by kevinasdzine** [reply]
It does look like some of the other features in OC are implemented with object types that are not ContentItem types but still stored in the Documents table (e.g. workflows) and query via ISession. For example, from the WorkflowController.cs

```
var workflowsQuery = _session.Query<Workflow, WorkflowIndex>(item => item.WorkflowId.IsIn(workflowIds));
```

I've tried to duplicate that type of object but it hasn't worked completely (writes entries into Documents but doesn't query).

This has been an _interesting_ discovery process.

**Reply by weirdyang** [reply]
> I'm having a little trouble prototyping a "vote" for a content item. I think I just need to define an object with the ContentItemId and a UserId, create an index provider and then use `SchemaBuilder.CreateMapIndexTableAsync` in my migration but it's not working quite like I'm expecting. Can you point me to any example code?

https://github.com/weirdyang/Cms.Web/blob/develop/Cms.Module.Api/TestObject/TestEntity.cs
